OneSignalSDK/onesignal/core/src/main/java/com/onesignal/internal/OneSignalImp.kt

```diff
@@ -333,33 +324,73 @@ internal class OneSignalImp(
-        if (!completed) {
-            throw IllegalStateException("initWithContext was not called or timed out")
-        }
-    }
-
     /**
      * Notifies both blocking and suspend callers that initialization is complete
      */
     private fun notifyInitComplete() {
         initAwaiter.complete()
     }
 
-    private suspend fun suspendUntilInit() {
+    /**
+     * Blocking version that waits for initialization to complete.
+     * Uses runBlocking to bridge to the suspend implementation.
+     * Preserves context-aware timeout behavior (shorter on main thread to prevent ANRs).
+     *
+     * @param timeoutMs Optional timeout in milliseconds. If not provided, uses context-aware timeout.
+     * @param operationName Optional operation name to include in error messages (e.g., "login", "logout")
+     */
+    private fun waitForInit(timeoutMs: Long? = null, operationName: String? = null) {
+        val actualTimeout = timeoutMs ?: initAwaiter.getDefaultTimeout()
+        runBlocking(ioDispatcher) {
+            waitUntilInitInternal(actualTimeout, operationName)
+        }
+    }
+
+    /**
+     * Suspend version that waits for initialization to complete.
+     * Uses context-aware timeout (shorter on main thread to prevent ANRs).
+     *
+     * @param timeoutMs Optional timeout in milliseconds. If not provided, uses context-aware timeout.
+     * @param operationName Optional operation name to include in error messages (e.g., "login", "logout")
+     */
+    private suspend fun suspendUntilInit(timeoutMs: Long? = null, operationName: String? = null) {
+        val actualTimeout = timeoutMs ?: initAwaiter.getDefaultTimeout()
+        waitUntilInitInternal(actualTimeout, operationName)
+    }
+
+    /**
+     * Common implementation for waiting until initialization completes.
+     * Handles all state checks and timeout logic.
+     *
+     * @param timeoutMs Timeout in milliseconds
+     * @param operationName Optional operation name to include in error messages (e.g., "login", "logout")
+     */
+    private suspend fun waitUntilInitInternal(timeoutMs: Long, operationName: String? = null) {
         when (initState) {
             InitState.NOT_STARTED -> {
-                throw IllegalStateException("Must call 'initWithContext' before use")
+                val message = if (operationName != null) {
+                    "Must call 'initWithContext' before '$operationName'"
+                } else {
+                    "Must call 'initWithContext' before use"
+                }
+                throw IllegalStateException(message)
             }
             InitState.IN_PROGRESS -> {
-                Logging.debug("Suspend waiting for init to complete...")
+                Logging.debug("Waiting for init to complete...")
                 try {
-                    withTimeout(MAX_TIMEOUT_TO_INIT) {
+                    withTimeout(timeoutMs) {
                         initAwaiter.awaitSuspend()
                     }
```

Comment:
If we timeout then have an inconsistent state where waitUntilInitInternal returns, but initState == InitState.IN_PROGRESS. I would rather we just wait here as long as it takes so code doesn't have to consider both states. I don't see what upsides this complexity gives us.

PR #2412 does this, and only logs if it is taking longer than expected.